### ISSUE TITLE
doxygen runtime went from 10minutes to >14hours (lots of `using namespace`)

### ISSUE THREAD

**Issue by martinus** [doxygen/doxygen#9266]
**Describe the bug**
Newest doxygen version is impracticably slow when there are lots of `using namespace` in the code. I stopped generating after 14 hours. Older versions take 10 minutes for the same codebase.

**Expected behavior**
Don't take forever

**To Reproduce**
Unfortunately I cannot share our code. It might be reproducible with several files and lots of namespaces, I can try to do that if you think that helps

**Version**
I checked out the latest version, and did a `git bisect` to find out when the slowdown was introduced. It first happened in this commit: 8a93f4e00b6922526935909312dabbd9a566d5cc. Before this commit a stripped down codebase takes 0.5 seconds to generate, after that commit I stopped it after several minutes. I did all my tests in Linux, with clang++

**Stack trace**
I used [hotspot](https://github.com/KDAB/hotspot) to figure out where doxygen was hanging (for this I built `master`), and got this flamegraph:
![image](https://user-images.githubusercontent.com/14386/162568380-fe6c0ad5-3be1-4572-bed3-5fffb5d93956.png)

**Additional context**
Looking at the changes in 8a93f4e00b6922526935909312dabbd9a566d5cc I suspect the issue is now too often a fresh `AccessStack` or `VisitedNamespaces` or `StringUnorderedSet` is used when previously a non-empty container was used

**Comment by albert-github** [comment]
You pointed to a probable cause of the problem (thanks), though you also wrote:
> Newest doxygen version is impracticably slow when there are lots of using namespace in the code. I stopped generating after 14 hours. Older versions take 10 minutes for the same codebase.

- Which is the "Newest doxygen version" version you tried (output of `doxygen -v`)
- Which is the most recent "Older versions" you tried?

especially the "Newest doxygen version" is of important.

The source code you used:
- which language is it written in?
- is it public domain or propriety?

**Comment by martinus** [comment]
> Which is the "Newest doxygen version" version you tried (output of doxygen -v)

I've built doxygen myself, this is the output of `doxygen -v` of the newest version that I've tried:

1.9.4 (de4a15d48dff81693c60043af52974cec7926c79)

> Which is the most recent "Older versions" you tried?

The most recent old version (last good version) I tried:

1.9.0 (02cdddb491739a18413025ddd587879bd8f5305f)

Then, first bad version is this:

1.9.0 (8a93f4e00b6922526935909312dabbd9a566d5cc)

**Comment by doxygen** [comment]
@martinus It would be very useful to have a dummy example, modelled after the proprietary code if needed, that allows me to reproduce the problem. It is enough if it shows a significant (say >2x) performance degradation.

**Comment by martinus** [comment]
I'm trying to create a stripped down example, but it seems to be a bit difficult to narrow it down. I see what I can do

**Comment by ferdymercury** [comment]
Not sure if related https://github.com/doxygen/doxygen/issues/5875

**Comment by martinus** [comment]
Sorry, it took me a long time to create a stripped down example that demonstrates slowness. In the end I have now this repository, which basically 2 files that have lots of `using namespace`: https://github.com/martinus/doxygen_slow_namespace_example

Generating doxygen for the ~80 lines of C++ code takes 1m 34s on my machine. It seems thow that the old version 1.9.0 is also slow on these files, the binary from 02cdddb491739a18413025ddd587879bd8f5305f takes 47 seconds.

**Comment by albert-github** [comment]
Thanks for the example, but why a repository (which is potential non persistent) and nut just a simple tar or zip file like: [example.tar.gz](https://github.com/doxygen/doxygen/files/8499921/example.tar.gz)

**Comment by albert-github** [comment]
It looks like it is related to the fixed problem #6382 but apparently we have her another case.

I ran with `doxygen -d time` (and set `QUIET=NO`) and see as potential place:
```
0.073 sec: Generating file sources...
0.074 sec: Generating code for file src/1.h...
531.519 sec: Generating file documentation...
```

**Comment by doxygen** [comment]
@martinus Thanks for the example, it should definitely not slow down things so much, so I'll look into it. That said, I see it uses "using namespace" in a header. Did you know this is considered a [bad practice](https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers)?

**Comment by martinus** [comment]
@doxygen Thank you for looking at this! You are right, `using namespace` in header is definitely bad practice, that code example I posted is not an indication of my programming style :laughing: It's just what I got from trying to reduce a large codebase down to a simple example to demonstrate the issue

**Comment by doxygen** [comment]
@martinus Please verify if the referenced commit fixes the problem.

**Comment by martinus** [comment]
@doxygen Fantastic, it works blazing fast! I've tried it on my stripped down example and the original input, and both are now very fast.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
